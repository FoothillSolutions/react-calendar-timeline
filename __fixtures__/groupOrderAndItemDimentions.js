export const orderedGroups = {
  '1': {
    group: {
      id: '1'
    },
    index: 0
  },
  '2': {
    group: {
      id: '2'
    },
    index: 1
  }
}

export const dimensionItems = [
  {
    dimensions: {
      collisionLeft: 1540540000000,
      collisionWidth: 6803877,
      height: 22.5,
      left: 4347.222222222223,
      order: {
        group: {
          id: '1'
        },
        index: 0
      },
      stack: true,
      top: 7.5,
      width: 236.24572916666668
    },
    id: '0'
  },
  {
    dimensions: {
      collisionLeft: 1540570000000,
      collisionWidth: 14875919,
      height: 22.5,
      left: 5388.888888888889,
      order: {
        group: {
          id: '1'
        },
        index: 0
      },
      stack: true,
      top: 7.5,
      width: 516.5249652777778
    },
    id: '1'
  },
  {
    dimensions: {
      collisionLeft: 1540620000000,
      collisionWidth: 20397548,
      height: 22.5,
      left: 7125,
      order: {
        group: {
          id: '1'
        },
        index: 0
      },
      stack: true,
      top: 7.5,
      width: 708.2481944444445
    },
    id: '2'
  }
]
