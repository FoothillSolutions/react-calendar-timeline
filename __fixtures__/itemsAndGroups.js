import moment from 'moment'

export const items = [
  {
    id: '0',
    group: '1',
    start_time: moment('2018-10-26T10:46:40.000').valueOf(),
    end_time: moment('2018-10-26T12:40:03.877').valueOf(),
    canMove: false,
    canResize: false
  },
  {
    id: '5',
    group: '1',
    start_time: moment('2018-10-26T08:46:40.000').valueOf(),
    end_time: moment('2018-10-26T14:40:03.877').valueOf(),
    canMove: false,
    canResize: false,
    className: ''
  },
  {
    id: '6',
    group: '1',
    start_time: moment('2018-10-26T13:46:40.000').valueOf(),
    end_time: moment('2018-10-26T20:40:03.877').valueOf(),
    canMove: false,
    canResize: false,
    className: ''
  },
  {
    id: '1',
    group: '1',
    start_time: moment('2018-10-26T19:06:40.000').valueOf(),
    end_time: moment('2018-10-26T23:14:35.919').valueOf(),
    canMove: true,
    canResize: 'both'
  },
  {
    id: '2',
    group: '1',
    start_time: moment('2018-10-27T08:00:00.000').valueOf(),
    end_time: moment('2018-10-27T13:39:57.548').valueOf(),
    canMove: false,
    canResize: false,
    className: ''
  },
  {
    id: '3',
    group: '3',
    start_time: moment('2018-10-27T18:00:00.000').valueOf(),
    end_time: moment('2018-10-27T23:39:57.548').valueOf(),
    canMove: false,
    canResize: false,
    className: ''
  }
]

export const groups = [{ id: '1' }, { id: '2' }, { id: '3' }]
