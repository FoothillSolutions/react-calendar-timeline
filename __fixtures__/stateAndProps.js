import { defaultKeys } from 'lib/default-config'
import moment from 'moment'

export const props = {
  keys: defaultKeys,
  lineHeight: 30,
  stackItems: true,
  itemHeightRatio: 0.75
}

export const visibleTimeStart = moment('2018-10-26T00:00:00.000')
export const visibleTimeEnd = moment('2018-10-27T00:00:00.000')

export const state = {
  draggingItem: undefined,
  dragTime: null,
  resizingItem: null,
  resizingEdge: null,
  resizeTime: null,
  newGroupOrder: null,
  canvasTimeStart: moment('2018-10-25T00:00:00.000').valueOf(),
  visibleTimeEnd: visibleTimeEnd.valueOf(),
  visibleTimeStart: visibleTimeStart.valueOf()
}
