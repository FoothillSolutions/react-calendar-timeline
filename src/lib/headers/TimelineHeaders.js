import React from 'react'
import { TimelineHeadersConsumer } from './HeadersContext'
import PropTypes from 'prop-types'
import SidebarHeader from './SidebarHeader'
import { RIGHT_VARIANT, LEFT_VARIANT } from './constants'
class TimelineHeaders extends React.Component {
  static propTypes = {
    registerScroll: PropTypes.func.isRequired,
    leftSidebarWidth: PropTypes.number.isRequired,
    rightSidebarWidth: PropTypes.number.isRequired,
    style: PropTypes.object,
    className: PropTypes.string,
    calendarHeaderStyle: PropTypes.object,
    calendarHeaderClassName: PropTypes.string
  }

  constructor(props) {
    super(props)
  }

  getRootStyle = () => {
    return {
      background: '#c52020',
      borderBottom: '1px solid #bbb',
      ...this.props.style,
      display: 'flex',
      width: '100%'
    }
  }

  getCalendarHeaderStyle = () => {
    const {
      leftSidebarWidth,
      rightSidebarWidth,
      calendarHeaderStyle
    } = this.props
    return {
      border: '1px solid #bbb',
      ...calendarHeaderStyle,
      overflow: 'hidden',
      width: `calc(100% - ${leftSidebarWidth + rightSidebarWidth}px)`
    }
  }

  render() {
    let rightSidebarHeader
    let leftSidebarHeader
    let calendarHeaders = []
    const children = Array.isArray(this.props.children)
      ? this.props.children.filter(c => c)
      : [this.props.children]
    React.Children.map(children, child => {
      if (
        child.type === SidebarHeader &&
        child.props.variant === RIGHT_VARIANT
      ) {
        rightSidebarHeader = child
      } else if (
        child.type === SidebarHeader &&
        child.props.variant === LEFT_VARIANT
      ) {
        leftSidebarHeader = child
      } else {
        calendarHeaders.push(child)
      }
    })
    return (
      <div style={this.getRootStyle()} className={this.props.className}>
        {leftSidebarHeader}
        <div
          ref={this.props.registerScroll}
          style={this.getCalendarHeaderStyle()}
          className={this.props.calendarHeaderClassName}
          data-testid="headerContainer"
        >
          {calendarHeaders}
        </div>
        {rightSidebarHeader}
      </div>
    )
  }
}

const TimelineHeadersWrapper = ({
  children,
  style,
  className,
  calendarHeaderStyle,
  calendarHeaderClassName
}) => (
  <TimelineHeadersConsumer>
    {({ leftSidebarWidth, rightSidebarWidth, registerScroll }) => {
      return (
        <TimelineHeaders
          leftSidebarWidth={leftSidebarWidth}
          rightSidebarWidth={rightSidebarWidth}
          registerScroll={registerScroll}
          children={children}
          style={style}
          className={className}
          calendarHeaderStyle={calendarHeaderStyle}
          calendarHeaderClassName={calendarHeaderClassName}
        />
      )
    }}
  </TimelineHeadersConsumer>
)

TimelineHeadersWrapper.propTypes = {
  style: PropTypes.object,
  className: PropTypes.string,
  calendarHeaderStyle: PropTypes.object,
  calendarHeaderClassName: PropTypes.string
}

export default TimelineHeadersWrapper
