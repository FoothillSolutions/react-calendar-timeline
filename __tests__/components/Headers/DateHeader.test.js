import React from 'react'
import { render, cleanup, within } from 'react-testing-library'
import Timeline from 'lib/Timeline'
import DateHeader from 'lib/headers/DateHeader'
import SidebarHeader from 'lib/headers/SidebarHeader'
import TimelineHeaders from 'lib/headers/TimelineHeaders'
import 'jest-dom/extend-expect'
import moment from 'moment'
import { items, groups } from '../../../__fixtures__/itemsAndGroups'
import { visibleTimeEnd, visibleTimeStart } from '../../../__fixtures__/stateAndProps'
const defaultProps = {
  groups,
  items,
  defaultTimeStart: moment(visibleTimeStart, 'x'),
  defaultTimeEnd: moment(visibleTimeEnd, 'x'),
}

describe("Testing DateHeader Component", () => {
  beforeEach(() => {

    Element.prototype.getBoundingClientRect = jest.fn(() => {
      return {
        width: 1000,
        height: 120,
        top: 0,
        left: 0,
        bottom: 0,
        right: 0,
      }
    });
  })
  afterEach(cleanup)

  // Testing The Example In The Docs
  it("Given DateHeader When rendered Then it should be rendered correctly in the timeLine", () => {
    const { getAllByTestId } = render(
      <Timeline
        {...defaultProps}
      >
        <TimelineHeaders>
          <SidebarHeader>
            {({ getRootProps }) => {
              return <div {...getRootProps()}>Left</div>
            }}
          </SidebarHeader>
          <DateHeader primaryHeader />
          <DateHeader />
          <DateHeader
            unit="day"
            style={{ height: 50 }}
            intervalRenderer={({ getIntervalProps, intervalContext }) => {
              return <div {...getIntervalProps()}>
                {intervalContext.intervalText}
              </div>
            }}
          />
        </TimelineHeaders>
      </Timeline>
    )

    expect(getAllByTestId('dateHeader')).toHaveLength(3)

  })

  it("Given Dateheader When pass a string typed labelFormat Then it should render the intervals with hte given format", () => {
    const { getAllByTestId } = renderDateHeader({ unit: "day", labelFormat: "MM/DD" });
    expect(getAllByTestId('dateHeader')[1]).toHaveTextContent('10/25')
    expect(getAllByTestId('dateHeader')[1]).toHaveTextContent('10/26')
    expect(getAllByTestId('dateHeader')[1]).toHaveTextContent('10/27')
    expect(getAllByTestId('dateHeader')[0]).toHaveTextContent('Thursday, October 25')

  })

  it("Given Dateheader When pass an object typed labelFormat Then it should render the intervals with hte given format", () => {
    const { getAllByTestId } = renderDateHeader({ unit: "day", labelFormat: { day: { long: "MM/DD/YYYY" } } });

    expect(getAllByTestId('dateHeader')[1]).toHaveTextContent('10/25/2018')
    expect(getAllByTestId('dateHeader')[1]).toHaveTextContent('10/26/2018')
    expect(getAllByTestId('dateHeader')[1]).toHaveTextContent('10/27/2018')
    expect(getAllByTestId('dateHeader')[0]).toHaveTextContent('Thursday, October 25')
  })
  it("Given Dateheader When pass a function typed labelFormat Then it should render the intervals with hte given format", () => {
    const formatlabel = jest.fn((interval, unit, labelWidth) => interval[0].format("MM/DD/YYYY"))
    const { getAllByTestId } = renderDateHeader({ unit: "day", labelFormat: formatlabel });

    expect(formatlabel).toHaveBeenCalled()

    expect(getAllByTestId('dateHeader')[1]).toHaveTextContent('10/25/2018')
    expect(getAllByTestId('dateHeader')[1]).toHaveTextContent('10/26/2018')
    expect(getAllByTestId('dateHeader')[1]).toHaveTextContent('10/27/2018')

    expect(getAllByTestId('dateHeader')[0]).toHaveTextContent('Thursday, October 25')

  })
  it("Given Dateheader When pass a string typed labelFormat Then it should be called with the right params", () => {
    const formatlabel = jest.fn((interval, unit, labelWidth) => interval[0].format("MM/DD/YYYY"))
    renderDateHeader({ unit: "day", labelFormat: formatlabel })
    expect(formatlabel).toHaveBeenCalled()
    expect(formatlabel).toHaveBeenCalledWith(expect.any(Array), "day", expect.any(Number))
  })


  it("Given Dateheader When click on the primary header Then it should change the format", async () => {

    const formatlabel = jest.fn((interval, unit, labelWidth) => interval[0].format("MM/DD/YYYY"))
    const { getByTestId, getAllByTestId } = renderDateHeader({ unit: "day", labelFormat: formatlabel });
    // Arrange
    const primaryHeader = getByTestId('dateHeader')
    const seconderyHeader = getAllByTestId('dateHeader')[2]
    const unitHeader = getAllByTestId('dateHeader')[1]

    const unitHeaderChildrenBeforeClick = unitHeader.childElementCount
    const secondaryHeaderChildrenBeforeClick = seconderyHeader.childElementCount

    // Act
    const primaryFirstClick = within(primaryHeader).getByText('Friday, October 26, 2018').parentElement
    primaryFirstClick.click()
    const primarySecondClick = within(primaryHeader).getByText('October 2018').parentElement
    primarySecondClick.click()

    // Assert
    expect(seconderyHeader).toHaveTextContent('January')
    expect(primaryHeader).toHaveTextContent('2018')
    expect(secondaryHeaderChildrenBeforeClick).toBeGreaterThan(seconderyHeader.childElementCount)
    expect(unitHeaderChildrenBeforeClick).toBeLessThan(unitHeader.childElementCount)

  })

  it("Given Dateheader When click on the secondary header Then it should change the format", async () => {

    const formatlabel = jest.fn((interval, unit, labelWidth) => interval[0].format("MM/DD/YYYY"))
    const { getByTestId, getAllByTestId } = renderDateHeader({ unit: "day", labelFormat: formatlabel });
    // Arrange
    const primaryHeader = getByTestId('dateHeader')
    const seconderyHeader = getAllByTestId('dateHeader')[2]
    const unitHeader = getAllByTestId('dateHeader')[1]

    const primaryFirstClick = within(primaryHeader).getByText('Friday, October 26, 2018').parentElement
    primaryFirstClick.click()
    const primarySecondClick = within(primaryHeader).getByText('October 2018').parentElement

    primarySecondClick.click()
    const unitHeaderChildrenBeforeClick = unitHeader.childElementCount
    const secondaryHeaderChildrenBeforeClick = seconderyHeader.childElementCount

    // Act
    const secondaryFirstClick = within(seconderyHeader).queryByText('January')
    secondaryFirstClick.click()
    const secondarySecondClick = within(seconderyHeader).queryByText('1')
    secondarySecondClick.click()

    // Assert
    expect(secondaryHeaderChildrenBeforeClick).toBeLessThan(seconderyHeader.childElementCount)
    expect(unitHeaderChildrenBeforeClick).toBeGreaterThan(unitHeader.childElementCount)
  })
  it('Given Dateheader When pass a className Then it should be applied to DateHeader', () => {
    const { getAllByTestId } = renderDateHeader({ labelFormat: "MM/DD/YYYY", className: 'test-class-name' });
    expect(getAllByTestId('dateHeader')[1]).toHaveClass('test-class-name')
  })
  it('Given Interval When pass an ovveride values for (width, left, position) it should not ovverride the default values', () => {
    const { getAllByTestId } = renderDateHeader({ labelFormat: "MM/DD/YYYY", props: { style: { width: 100, position: 'fixed' } } });
    const { width, position } = getComputedStyle(getAllByTestId('interval')[0])
    expect(width).not.toBe('100px')
    expect(position).not.toBe('fixed')


  })

  it('Given Interval When pass an override (width, position) Then it should render the default values for it', () => {
    const { getAllByTestId } = renderDateHeader({ labelFormat: "MM/DD/YYYY", props: { style: { width: 100, position: 'fixed' } } });
    const { width, position } = getComputedStyle(getAllByTestId('interval')[0])
    expect(width).toBe('36px')
    expect(position).toBe('absolute')


  })
  it('Given Interval When pass any style other than (position, width, left) through the Dateheader Then it should take it', () => {
    const { getAllByTestId } = renderDateHeader({ labelFormat: "MM/DD/YYYY", props: { style: { display: 'flex' } } });
    const { display } = getComputedStyle(getAllByTestId('interval')[0])

    expect(display).toBe('flex')

  })
  it('Given unit Dateheader When pass a style Object Then it should render the given style correctly', () => {
    const { getAllByTestId } = renderDateHeader({ style: { height: 50 }, labelFormat: "MM/DD/YYYY" });
    const { height } = getComputedStyle(getAllByTestId('dateHeader')[1])

    expect(height).toBe('50px')
  })

})

function renderDateHeader({ labelFormat, unit, props, className, style } = {}) {

  return render(

    <Timeline
      {...defaultProps}
      timelineUnit="month"
      style={{ width: 1000 }}
    >
      <TimelineHeaders>
        <SidebarHeader>
          {({ getRootProps }) => {
            return <div {...getRootProps()}>Left</div>
          }}
        </SidebarHeader>
        <DateHeader primaryHeader />
        <DateHeader
          unit={unit}
          labelFormat={labelFormat}
          props={props}
          style={style}
          className={className}
          intervalRenderer={({ getIntervalProps, intervalContext }, props) => {
            return <div data-testid="interval" {...getIntervalProps(props)}>
              {intervalContext.intervalText}

            </div>
          }}
        />
        <DateHeader />
      </TimelineHeaders>
    </Timeline>
  )
}