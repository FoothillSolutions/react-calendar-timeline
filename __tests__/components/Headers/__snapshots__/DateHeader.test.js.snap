// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing DateHeader Component DateHeader Unit Values Given DateHeader When not passing a unit then the date header unit should be same as timeline unit 1`] = `
Array [
  "10/24/2018 12:00 am",
  "10/25/2018 12:00 am",
  "10/26/2018 12:00 am",
  "10/27/2018 12:00 am",
]
`;

exports[`Testing DateHeader Component DateHeader Unit Values Given DateHeader When passing a unit then the date header unit should be same as unit passed 1`] = `
Array [
  "10/24/2018 11:00 pm",
  "10/25/2018 12:00 am",
  "10/25/2018 01:00 am",
  "10/25/2018 02:00 am",
  "10/25/2018 03:00 am",
  "10/25/2018 04:00 am",
  "10/25/2018 05:00 am",
  "10/25/2018 06:00 am",
  "10/25/2018 07:00 am",
  "10/25/2018 08:00 am",
  "10/25/2018 09:00 am",
  "10/25/2018 10:00 am",
  "10/25/2018 11:00 am",
  "10/25/2018 12:00 pm",
  "10/25/2018 01:00 pm",
  "10/25/2018 02:00 pm",
  "10/25/2018 03:00 pm",
  "10/25/2018 04:00 pm",
  "10/25/2018 05:00 pm",
  "10/25/2018 06:00 pm",
  "10/25/2018 07:00 pm",
  "10/25/2018 08:00 pm",
  "10/25/2018 09:00 pm",
  "10/25/2018 10:00 pm",
  "10/25/2018 11:00 pm",
  "10/26/2018 12:00 am",
  "10/26/2018 01:00 am",
  "10/26/2018 02:00 am",
  "10/26/2018 03:00 am",
  "10/26/2018 04:00 am",
  "10/26/2018 05:00 am",
  "10/26/2018 06:00 am",
  "10/26/2018 07:00 am",
  "10/26/2018 08:00 am",
  "10/26/2018 09:00 am",
  "10/26/2018 10:00 am",
  "10/26/2018 11:00 am",
  "10/26/2018 12:00 pm",
  "10/26/2018 01:00 pm",
  "10/26/2018 02:00 pm",
  "10/26/2018 03:00 pm",
  "10/26/2018 04:00 pm",
  "10/26/2018 05:00 pm",
  "10/26/2018 06:00 pm",
  "10/26/2018 07:00 pm",
  "10/26/2018 08:00 pm",
  "10/26/2018 09:00 pm",
  "10/26/2018 10:00 pm",
  "10/26/2018 11:00 pm",
  "10/27/2018 12:00 am",
  "10/27/2018 01:00 am",
  "10/27/2018 02:00 am",
  "10/27/2018 03:00 am",
  "10/27/2018 04:00 am",
  "10/27/2018 05:00 am",
  "10/27/2018 06:00 am",
  "10/27/2018 07:00 am",
  "10/27/2018 08:00 am",
  "10/27/2018 09:00 am",
  "10/27/2018 10:00 am",
  "10/27/2018 11:00 am",
  "10/27/2018 12:00 pm",
  "10/27/2018 01:00 pm",
  "10/27/2018 02:00 pm",
  "10/27/2018 03:00 pm",
  "10/27/2018 04:00 pm",
  "10/27/2018 05:00 pm",
  "10/27/2018 06:00 pm",
  "10/27/2018 07:00 pm",
  "10/27/2018 08:00 pm",
  "10/27/2018 09:00 pm",
  "10/27/2018 10:00 pm",
]
`;

exports[`Testing DateHeader Component DateHeader Unit Values Given DateHeader When passing primaryHeader Then the header unit should be bigger the timeline unit 1`] = `
Array [
  "10/01/2018 12:00 am",
]
`;

exports[`Testing DateHeader Component DateHeader Unit Values Given DateHeader When passing secondaryHeader Then the header unit should be same as the timeline unit 1`] = `
Array [
  "10/24/2018 12:00 am",
  "10/25/2018 12:00 am",
  "10/26/2018 12:00 am",
  "10/27/2018 12:00 am",
]
`;
